mediorum/ddl/ddl.go

```diff
@@ -71,6 +82,133 @@ func md5string(s string) string {
+	var partitioned bool
+	db.QueryRow(`select count(*) = 1 from mediorum_migrations where hash = $1`, partition_ops_completed).Scan(&partitioned)
+	if partitioned {
+		slog.Info("already partitioned ops, skipping migration scheduling")
+	} else {
+		slog.Info("scheduling partition ops migration", "time", randomTime)
+		time.AfterFunc(randomTime, func() {
+			slog.Info("marking migrations table to partition ops after we restart...")
+			mustExec(db, `insert into mediorum_migrations values ($1, now()) on conflict do nothing`, partition_ops_scheduled)
+			os.Exit(0)
+		})
+	}
+}
+
+func migratePartitionOps(db *sql.DB) {
+	slog.Info("checking if it's time to partition ops...")
+	var scheduled bool
+	db.QueryRow(`select count(*) = 1 from mediorum_migrations where hash = $1`, partition_ops_scheduled).Scan(&scheduled)
+	if !scheduled {
+		fmt.Println("partition ops migration not scheduled, skipping ddl")
+		return
+	}
+
+	logfileName := "partition_ops.txt"
+
+	logAndWriteToFile(fmt.Sprint("starting partitioning of ops"), logfileName)
+	start := time.Now()
+
+	mustExec(
+		db,
+		`begin;
+		-- Kill all long-running sql queries
+		SELECT pg_terminate_backend(pg_stat_activity.pid) FROM pg_stat_activity WHERE state = 'active' AND now() - query_start > interval '1 minute';
+
+    -- Rename ops to old_ops if this has not already been done
+		DO $$ 
+		BEGIN
+				IF NOT EXISTS (SELECT 1 FROM information_schema.tables WHERE table_name = 'old_ops') THEN
+						ALTER TABLE IF EXISTS ops RENAME TO old_ops;
+				END IF;
+		END $$;
+
+		-- Create the new ops parent table
+		CREATE TABLE IF NOT EXISTS ops (
+			"ulid" TEXT,
+			"host" TEXT,
+			"action" TEXT,
+			"table" TEXT,
+			"data" JSONB,
+			"transient" BOOLEAN
+		) PARTITION BY HASH ("host");
+
+		DO $$ 
+		DECLARE 
+			i INTEGER;
+			partition_name TEXT;
+		BEGIN 
+			FOR i IN 0..1008 LOOP -- 1009 partitions
+				partition_name := 'ops_' || i;
+				EXECUTE 'CREATE TABLE IF NOT EXISTS' || partition_name || ' PARTITION OF ops FOR VALUES WITH (MODULUS 1009, REMAINDER ' || i || ');';
+				EXECUTE 'ALTER TABLE ' || partition_name || ' ADD PRIMARY KEY ("ulid");';
+			END LOOP; 
+		END $$;
+
+		commit;`,
+	)
+
+	err := migrateOpsData(db, logfileName)
+	if err != nil {
+		logAndWriteToFile(err.Error(), logfileName)
+		log.Fatal(err)
+	}
+
+	mustExec(
+		db,
+		`begin;
+		DROP TABLE old_ops;
+		insert into mediorum_migrations values ($1, now()) on conflict do nothing;
+		commit;`,
+		partition_ops_completed,
+	)
+
+	logAndWriteToFile(fmt.Sprintf("finished partitioning ops. took %gm\n", time.Since(start).Minutes()), logfileName)
+}
+
+// copy data from old_ops to ops
+func migrateOpsData(db *sql.DB, logfileName string) error {
+	logAndWriteToFile(fmt.Sprintln("starting ops data migration"), logfileName)
+	lastUlid := ""
+	pageSize := 1000
+	rowsMigrated := 0
+
+	for {
+		rows, err := db.Query(`SELECT * FROM old_ops WHERE ulid > $1 ORDER BY "ulid" ASC LIMIT $2`, lastUlid, pageSize)
```

Comment:
just want to double-check this won't cause any issues and ordering by descending with the cursors isn't necessary? maybe this way works well because it's not using host so it's just a simple sequential scan with the index for ulids that are all right next to each other?

Reply:
need the ordering so that the last record in the batch is the correct value for `lastUlid` for the next iteration

tested and explain analyzed the sql, works fine and uses the ulid index so its actually an index scan